src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -939,3 +957,222 @@ void AppHost::_SummonWindowRequested(const winrt::Windows::Foundation::IInspecta
+    if (hmenu)
+    {
+        MENUINFO mi{};
+        mi.cbSize = sizeof(MENUINFO);
+        mi.fMask = MIM_STYLE | MIM_APPLYTOSUBMENUS | MIM_MENUDATA;
+        mi.dwStyle = MNS_NOTIFYBYPOS;
+        mi.dwMenuData = NULL;
+        SetMenuInfo(hmenu, &mi);
+
+        // Focus Current Terminal Window
+        AppendMenu(hmenu, MF_STRING, (UINT_PTR)TrayMenuItemAction::FocusTerminal, RS_(L"TrayIconFocusTerminal").c_str());
+        AppendMenu(hmenu, MF_SEPARATOR, 0, L"");
+
+        // Submenu for Windows
+        if (auto windowSubmenu = _CreateWindowSubmenu())
+        {
+            MENUINFO submenuInfo{};
+            submenuInfo.cbSize = sizeof(MENUINFO);
+            submenuInfo.fMask = MIM_MENUDATA;
+            submenuInfo.dwStyle = MNS_NOTIFYBYPOS;
+            submenuInfo.dwMenuData = (UINT_PTR)TrayMenuItemAction::SummonWindow;
+            SetMenuInfo(windowSubmenu, &submenuInfo);
+
+            AppendMenu(hmenu, MF_POPUP, (UINT_PTR)windowSubmenu, RS_(L"TrayIconWindowSubmenu").c_str());
+        }
+    }
+    return hmenu;
+}
+
+// Method Description:
+// - Create a menu with a menu item for each window available to summon.
+//   If a window is unnamed, we'll use its ID but still mention that it's unnamed.
+// Arguments:
+// - <none>
+// Return Value:
+// - The handle to the newly created window submenu.
+HMENU AppHost::_CreateWindowSubmenu()
+{
+    if (auto hmenu = CreatePopupMenu())
+    {
+        auto locWindow = RS_(L"WindowIdLabel");
+        auto locUnnamed = RS_(L"UnnamedWindowName");
+        for (const auto [id, name] : _windowManager.GetPeasantNames())
+        {
+            winrt::hstring displayText = name;
+            if (name.empty())
+            {
+                displayText = fmt::format(L"{} {} - <{}>", locWindow, id, locUnnamed);
+            }
+
+            AppendMenu(hmenu, MF_STRING, (UINT_PTR)id, displayText.c_str());
+        }
+        return hmenu;
+    }
+    return nullptr;
+}
+
+// Method Description:
+// - This is the handler for when one of the menu items are selected within
+//   the tray icon's context menu.
+// Arguments:
+// - menu: The handle to the menu that holds the menu item that was selected.
+// - menuItemIndex: The index of the menu item within the given menu.
+// Return Value:
+// - <none>
+void AppHost::_TrayMenuItemSelected(const HMENU menu, const UINT menuItemIndex)
+{
+    // Let's find out which menu/submenu we're looking at.
+    MENUINFO mi{};
+    mi.cbSize = sizeof(MENUINFO);
+    mi.fMask = MIM_MENUDATA;
+    GetMenuInfo(menu, &mi);
+
+    if (mi.dwMenuData)
+    {
+        if ((TrayMenuItemAction)mi.dwMenuData == TrayMenuItemAction::SummonWindow)
+        {
+            Remoting::SummonWindowSelectionArgs args{};
+            args.WindowID(GetMenuItemID(menu, menuItemIndex));
+            args.SummonBehavior().ToggleVisibility(false);
+            _windowManager.SummonWindow(args);
+        }
+    }
+
+    auto action = (TrayMenuItemAction)GetMenuItemID(menu, menuItemIndex);

```

Comment:
`const` and another c-style cast

Reply:
huh, if you replace `auto` with `TrayMenuItemAction`, you might not even need to manually cast? (not sure)

Reply:
Looks like I can't replace auto 😔